### DISCUSSION TITLE
Integrate HTML forms?

### DISCUSSION THREAD

**Discussion by AnthonyMuscio** [TiddlyWiki/TiddlyWiki5#5873] [Ideas]
Folks,

I suppose I know enough to be dangerous, and not enough to implement such things. I also have a deep knowledge of IT which means I often know when something is possible, but I do not know how to do it myself or explain all the details.

Why: because there is lots of free IP (intellectual property) out there if we could only integrate tiddlywiki with bog standard html even if it is limited to a few actions. I can argue this "why would do this" a lot more but for now I will assume your imagination can see where this could go.

Perhaps I miss understand but does the **event catcher widget** offer a method to achieve the following, or can you see a way to do this?

I understand tiddlywiki uses the DOM document object model differently but I wonder if we can interface with the standard DOM such as we could provision a little support to make use of available forms. As I understand it there is a level of sand boxing.

Please look at the following tiddlywiki https://anthonymuscio.github.io/testforms.html

- I have added to an empty.html only two tiddlers a "stylesheet and a form".
- You can see how the styles disrupt the wiki as a whole but it is mostly usable.
- I sourced this here https://www.w3docs.com/tools/editor/5887

Now as I understand it everything is working until the user hits submit and we have no backend to process it, but we do know everything is in the browser memory/dom. Until refreshed that data remains on the screen.

**So my Question is;**
Can anyone envision a way to provision A service in tiddlywiki that such a form could be interrogated and store the result in a tiddler for subsequent use? Ideally such a form could also be populated from data in a tiddlers fields as well to allow editing but just a submission equivalent would be very useful. Ideally this would be without modifying every input to a widget, but extracting content and saving it. Or perhaps a special parsing layer that converted html to widgets?

Past experiments;
Just so you are aware I have placed such html into a tiddler before with html, Javascript and php underlying the tiddlywiki in the same only host and seen that such forms can be operational. This is in itself a possible avenue to expose login and security and permit saving to a host not unlike tw-reciever does. However in this discussion I am hoping someone with more skills than I can find a way to make this a reality within TiddlyWiki.

Regards
Tones

**Comment by AnthonyMuscio** [comment]
Updated with public URL https://anthonymuscio.github.io/testforms.html

**Comment by pmario** [comment]
>Just so you are aware I have placed such html into a tiddler before with html, Javascript and php underlying the tiddlywiki in the same only host and seen that such forms can be operational. This is in itself a possible avenue to expose login and security and permit saving to a host not unlike tw-reciever does. However in this discussion I am hoping someone with more skills than I can find a way to make this a reality within TiddlyWiki.

I think this example doesn't work. The text elements don't have any IDs, classes or labels, that could be used to reliably store content in a tiddler. ... 

I would go a completely different way. .. Your example form is built with 3 different elements. 

 - text input 1 line
 - text input multi-line 
 - radio selector

So I would make 3 macros, that do the job and work with TW. IMO it will be 100 times less work, than building a general purpose parser that tries to make sense of a broken or incomplete html-forms copy-pasted from somewhere in the web. 

Having a look at: https://www.w3docs.com/learn-html/html-form-templates.html .... There may be about 20 different elements and that's it. 

Having macros, that put all the elements into a single dictionary tiddler and send the feedback as a feedback.tid file per mail, in a text-only mode, imo would go a long way. 

Using .tid as the file format, will avoid all the security and validation issues, that forms normally have.

**Reply by pmario** [reply]
There is 1 more thing: The example form is English only. .. What about the other 20 languages, that are supported by TW?

**Reply by AnthonyMuscio** [reply]
Mario,

Thanks for your feedback. I can see that the value of the "found" form is its css, layout and the text ie/questions and the choice of input elements. This information remains there. As you point out there are a limited set of form/input tags that we need to respond to.

The value of the Original Thread (OT) post approach would be far more extensive a solution, but as you suggest perhaps a more specific solution will do, for which arguable there would be somewhat less benefits.

So to account for all possible forms, you may suggest we find widget replacements for the html form elements we need to from the following set` <input> <textarea> <button> <select> <option> <fieldset> <label>` and most of these are already available. 

So what remains for us to find is a method to assist the designer translating such a form into a tiddlywiki form, that acts on tiddlywiki objects. The following methods in order of ease of use may be the approach;

1. The OT pulling the values in the form into tiddlywiki storage or via your suggested tid files see "via local storage below" as well
2. Get this to be parsed into tiddlywiki mark-up and widgets (your points taken into consideration on this one)
2. Have another tiddler parse this html form tiddler and build a tiddlywiki alternative that works
3. Have a set of tools in the editor to help convert this to widgets and text references (tiddler and fields) where needed.
4. Manually change all html to widgets as needed.

Via local storage?
Another though may be provide a method by which the submit stored the results in local storage and another tool can pull that from local storage and apply it to a tiddler, a bit like a copy and paste process. In this case there may need to be a submit equivalent widget that does this. The paste format perhaps in a data-tiddler or JSON format. In some ways the question becomes can we convert information in the DOM to a JSON.

Regards
Tony

**Reply by AnthonyMuscio** [reply]
Mario, I do not know how to handle multiple languages, but I imagine one would present a different form to a user in the other language, or enable google translate on top of it.

**Reply by AnthonyMuscio** [reply]
Mario; you said; _I think this example doesn't work_ but I think it does. this example "works" as html/php not to give tiddlywiki the values. This is something **I did in the past**, TiddlyWiki just becomes the presentation layer.

The Original thread suggests we can do better than how **I did this in the past**, with a desire to get data into a tiddler in the current tiddlywiki.

**Reply by pmario** [reply]
The problem we get with copy/paste forms sent to a server is "spam". 

So if you need a feedback form and you don't want to deal with the problems associated, you'll probably need a 3rd party service.

or

you'll need to invest some work, to do it the TW way.
